### PR TITLE
`Development`: Track locked status for programming exercise participations

### PR DESCRIPTION
**Pull request by frederik-kt**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
Logic for whether participations are locked due to the start date, the due date, or the submission limit, is scattered around the server and client code.

### Description
This PR adds a field `locked` to the  participation table that is kept up to date.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise without dates or submission limit configured, offline IDE should be allowed

1. Log in to Artemis as the student and participate in the exercise.
2. Log in as the instructor and edit the exercise, disallowing offline IDE usage.
3. Log in as the student again. Make sure you can still participate in the online editor but not by using your local Git client.
4. Log in as the instructor and edit the exercise again, e.g. by adding a due date in the past, or adding a submission limit (you cannot add the submission limit via the normal update exercise view, you need to go to the grading of the exercise for that)
5. Make sure the changes are reflected for the student. E.g. when you introduced a submission limit, make sure the student cannot submit after reaching that limit and "Repository is locked" is shown after reloading)

#### Exam Mode Testing
Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Programming Exercise with a submission limit configured

1. Log in to Artemis
2. Participate in the exam as a student and submit until hitting the submission limit
3. Make sure you receive an error message that informs you about the submission limit. Reload the page and make sure the online editor is locked.
4. After handing in the exam, make sure you cannot push via your local Git client anymore.

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exam-submission.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/exercises/programming/programming-exam-submission.component.ts.html) | 88.88% | ok |
| [code-editor-actions.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/code-editor/actions/code-editor-actions.component.ts.html) | 81.14% |ok |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ProgrammingExerciseStudentParticipation.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.participation/ProgrammingExerciseStudentParticipation.html) | 100% | ok |
| [ParticipationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ParticipationService.html) | 94% | ok |
| [RepositoryAccessService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/RepositoryAccessService.html) | 98% | ok |
| [SubmissionPolicyService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/SubmissionPolicyService.html) | 92% | ok |
| [ExamService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/ExamService.html) | 95% | ok |
| [InstanceMessageReceiveService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.messaging/InstanceMessageReceiveService.html) | 75% | ok |
| [MainInstanceMessageSendService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.messaging/MainInstanceMessageSendService.html) | 96% | ok |
| [ProgrammingExerciseGradingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseGradingService.html) | 97% | ok |
| [ProgrammingExerciseParticipationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseParticipationService.html) | 70% | ok, all changes covered |
| [ProgrammingExerciseRepositoryService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseRepositoryService.html) | 92% | ok |
| [ProgrammingExerciseScheduleService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.scheduled/ProgrammingExerciseScheduleService.html) | 86% | ok |
| [ParticipationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4834-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ParticipationResource.html) | 91% | ok |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/participation/ProgrammingExerciseStudentParticipation.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/domain/participation/ProgrammingExerciseStudentParticipation.java:34]:
```suggestion
    private boolean locked = true;
```
Can you make this a regular `boolean` here and add non-null and a default value to the database schema definition?

In some other places (e.g. ProgrammingExercise) we have gor nullable booleans where the `null` value is sometimes interpreted as `true`, sometimes as `false`, which of course has to be checked everywhere with e.g. `Booleans.TRUE.equals(...)`. This makes the code quite error-prone.

Comment (frederik-kt) [reply src/main/java/de/tum/in/www1/artemis/domain/participation/ProgrammingExerciseStudentParticipation.java]: Changed it to a regular `boolean`

**Review by JohannesStoehr — DISMISSED** [review]
Code changes look good to me. This should make life easier for us developers from now on 👍

**Review by JohannesStoehr — DISMISSED** [review]
New tests look good as well

**Review by undernagruzez — COMMENTED** [review]
Tested on ts3, the exam mode appeared to be functioning properly, but the submission count was set to 1 and did not update even after refreshing the page. The message stated I can still push, but I could not, as the limit had been reached already.
image
In exercise mode, I attempted to update the submission limit for a particular exercise but was still able to submit beyond the set limit. Additionally, I was unable to push my local repositories to the remote origin, which is expected behavior.

### Reply
Thanks for testing! The non-updating submission count is a known bug that is being fixed in another PR: https://github.com/ls1intum/Artemis/pull/6568.

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreScheduleService.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreScheduleService.java:122]: This change is unacceptable. We cannot merge the PR with it!

Comment (frederik-kt) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreScheduleService.java]: removed

**Review by tobias-lippert — APPROVED** [review]
Tested on ts2.
Both participation in course exercises and exam exercises wourked as expected.
The only aspect I noticed was that the count of the number of submissions is not shown in the code editor.
Hasn't this been introduced recently?

**Review by rriyaldhi — APPROVED** [review]
Tested the exam mode on TS4 using both the online editor and local git client. Works as expected.

**Review by pellumbbaboci — DISMISSED** [review]
Tested during testing session on ts4, works as expected

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS11 during testing session. Works as intended. I guess the error messages are quite confusing for students. Might be something for a followup pr

**Review by rriyaldhi — DISMISSED** [review]
Tested on TS11 and works as described.

**Review by laadvo — DISMISSED** [review]
Tested on ts11 during the testing session, the locked status works as expected

**Review by MaximilianAnzinger — DISMISSED** [review]
Tested on TS11 during testing session. Works as intended.

**Review by rriyaldhi — DISMISSED** [review]
Tested the exam mode on TS11 and works as described.
Screenshot 2023-06-09 at 15 17 24
Screenshot 2023-06-09 at 15 18 29

**Review by tobias-lippert — DISMISSED** [review]
Tested in testing session on ts11. Everything worked and submission limit is enforced.

**Review by dearjasmina — COMMENTED** [review]
Maybe I didn't do something correctly regarding submission policy, but I could submit more than supposed allowed, see photo:
image

## Reply
You need to activate the submission policy to have it take effect:
Screenshot (187)

However, I think we should not show the submission count if the policy is not activated. I created an issue for this: https://github.com/ls1intum/Artemis/issues/6719

**Review by tobias-lippert — DISMISSED** [review]
Reapprove, only test files were affected and the tests pass.

**Review by MaximilianAnzinger — APPROVED** [review]
Reapproved.

**Review by rriyaldhi — APPROVED** [review]
Reapprove after merge conflict

**Review by undernagruzez — DISMISSED** [review]
Reapproved

**Review by MaximilianAnzinger — DISMISSED** [review]
Reapproved after merge conflict.

**Review by pellumbbaboci — DISMISSED** [review]
re-approve after conflicts

**Review by tobias-lippert — APPROVED** [review]
Reapprove after fixing a merge conflict.

**Review by rriyaldhi — APPROVED** [review]
Reapprove after merge conflict

**Review by tobias-lippert — APPROVED** [review]
Reapprove after fixing merge conflict.

**Review by rriyaldhi — APPROVED** [review]
Reapprove

**Review by MaximilianAnzinger — APPROVED** [review]
Reapproved.

**Review by pellumbbaboci — APPROVED** [review]
re-approve after merge conflicts 👍

**Review by undernagruzez — APPROVED** [review]
Reapproved

**Review by dearjasmina — APPROVED** [review]
reapproved

**Pull request merged**
